### PR TITLE
Dev user endpoint

### PR DESCRIPTION
**Pull request by brngylni**
This PR adds the User endpoint #5

### CODE REVIEW

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from flask import Blueprint, Flask, request, Response, Request
+from ..database import db
+from ..models.user import User
+from sqlalchemy_helpers import get_or_create
+
+app = Flask(__name__)
+user_endpoint = Blueprint("user_endpoint", __name__)
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:8]: Please leave two lines of difference between each segregated block definition.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+@user_endpoint.route("/user", methods=["POST"])
+def create_user():
+    """
+        Used for creating a new user by sending a post request to /user/ path.
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:12]: Get rid of the added tab please.

```suggestion
    Used for creating a new user by sending a post request to /user/ path.
```

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    """
+    
+    if not validate_request(request):
+        return Response("{'message': 'Bad Request'}", status=400, mimetype='application/json')
+    session = db.Session()
+    user, created = get_or_create(session, User, username=request.json['username'])
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:18]: Add exception handling for `KeyError` (when the `username` key is not found in the JSON body). Abort the request with the `HTTP 429 Unprocessable Entity` error.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    if not created:
+        return Response("{'message': 'User Already Exists'}", status=409, mimetype='application/json')
+    else:
+        return Response("{'message': 'Created', 'uuid': '%s'}" % user['id'], status=201, mimetype='application/json')
+        
+    
+@user_endpoint.route("/user/search", methods=["GET"])
+def get_user():
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:27]: Use this empty line ;-)

If docstrings are created for a function of a kind, please ensure that they are created for other functions of that kind.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        return Response("{'message': 'Created', 'uuid': '%s'}" % user['id'], status=201, mimetype='application/json')
+        
+    
+@user_endpoint.route("/user/search", methods=["GET"])
+def get_user():
+    
+    if not validate_request(request):
+        return Response("{'message': 'Bad Request'}", status=400, mimetype='application/json')
+    session = db.Session()
+    users = session.query(User).filter(User.username.like(request.json['username'])).all()
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:31]: Add exception handling for `KeyError` (when the `username` key is not found in the JSON body). Abort the request with the `HTTP 429 Unprocessable Entity` error.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        return Response("{'user_list': "+ users +"}", status=200, mimetype='application/json')
+    
+    
+@user_endpoint.route("/user", methods=["GET"])
+def lookup_user():
+    if not validate_request(request):
+        return Response("{'message': 'Bad Request'}", status=400, mimetype='application/json')
+    session = db.Session()
+    
+    user = session.query(User).filter(User.username == request.json['username']).first()
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:44]: Add exception handling for `KeyError` (when the `username` key is not found in the JSON body). Abort the request with the `HTTP 429 Unprocessable Entity` error.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    if user is None:
+        return Response("{'message': 'Not Found'}", status=404, mimetype='application/json')
+    else:
+        return Response("{'uuid': "+ user.id +", 'username': '"+ user.username +"'}", status=200, mimetype='application/json')
+    
+    
+    
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:52]: More than two lines are left empty before the function definition.

Just two lines should be enough.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    session = db.Session()
+    users = session.query(User).filter(User.username.like(request.json['username'])).all()
+    if users is None or users == []:
+        return Response("{'message': 'Not Found'}", status=404, mimetype='application/json')
+    else:
+        return Response("{'user_list': "+ users +"}", status=200, mimetype='application/json')
+    
+    
+@user_endpoint.route("/user", methods=["GET"])
+def lookup_user():
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:39]: If docstrings are created for a function of a kind, please ensure that they are created for other functions of that kind.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+from flask import Blueprint, Flask, request, Response, Request
+from ..database import db
+from ..models.user import User
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:3]: Ref. https://github.com/fedora-infra/webhook-to-fedora-messaging/pull/24#discussion_r1667901072

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    
+    
+    
+    
+def validate_request(request: Request):
+    if request.json['username'] is None:
+        return False
+    return True
+    
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:58]: Leave one empty line at the end of the file.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    else:
+        return Response("{'uuid': "+ user.id +", 'username': '"+ user.username +"'}", status=200, mimetype='application/json')
+    
+    
+    
+    
+def validate_request(request: Request):
+    if request.json['username'] is None:
+        return False
+    return True
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:56]: Maybe convert this to a decorator?

File: webhook_to_fedora_messaging/main.py [hunk 12]
```diff
@@ -15,7 +15,7 @@
 from .config.defaults import LOGGER_CONFIG
 
 from logging.config import dictConfig
-
+from .endpoints.user import user_endpoint
```
Comment (gridhead) [inline webhook_to_fedora_messaging/main.py:18]: Ref. https://github.com/fedora-infra/webhook-to-fedora-messaging/pull/24#discussion_r1667901072

File: webhook_to_fedora_messaging/main.py [hunk 13]
```diff
@@ -26,7 +26,7 @@ def create_app():
     main = Flask(
 	    "webhook_to_fedora_messaging"
     )
-
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/main.py:29]: Leave the line empty.

```suggestion

```

**Review by gridhead — CHANGES_REQUESTED** [review]
Also, please follow https://github.com/fedora-infra/webhook-to-fedora-messaging/pull/2#issuecomment-2134623677.

File: webhook_to_fedora_messaging/endpoints/user.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+app = Flask(__name__)
+user_endpoint = Blueprint("user_endpoint", __name__)
+
+
+@user_endpoint.before_request
+def validation():
+    if not validate_request(request):
+        return unprocessable_entity()
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/user.py:15]: Please leave two lines of difference between each segregated block definition.

File: webhook_to_fedora_messaging/endpoints/util.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    return Response(data, status=201, mimetype='application/json')
+
+
+def conflict(data: dict) -> Response:
+    return Response(data, status=409, mimetype='application/json')
+
+
+def validate_request(request: dict, fields=['username']):
+    return all(field in request for field in fields)
+
```
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/util.py:26]: Please leave two lines of difference between each segregated block definition.

**Review by gridhead — CHANGES_REQUESTED** [review]
Please use the emojis to acknowledge the changes made and leave it for the reviewer to decide if the requested change has been made or not.

Once you are done with making the remaining changes - please squash the commits into one and then, it should be ready for a merge.

**Review by gridhead — CHANGES_REQUESTED** [review]
LGTM

@brngylni, please resolve the merge conflict and squash all commits into one.

**Comment by brngylni** [comment]
> LGTM
> 
> @brngylni, please resolve the merge conflict and squash all commits into one.

Done!

**Review by gridhead — APPROVED** [review]
I am not dragging this any longer than it has to be. Other required changes can be made to the codebase after it is merged.

Merging now.

**Pull request merged**